### DISCUSSION TITLE
Struggling to get kerning to match what SVG generates in browser...

### DISCUSSION THREAD

**Discussion by avisra** [SixLabors/Fonts#403] [Q&A] [answered]
I am using the following code in attempt to draw "PLAYER NAME" in the "Kelly Slab Regular" font (from google).

I need it to match an SVG preview which is being provided on the web. Here is what the SVG preview displays using CSS fonts
![svgexample](https://github.com/SixLabors/Fonts/assets/433538/fc57636f-1ccb-446a-bb64-5cf7221bf47d)

But when I try to generate this with SixLabors using the below code, I receive the following:
![image](https://github.com/SixLabors/Fonts/assets/433538/5fb07e45-7af5-49e8-84d7-b379ad81626d)
image.Mutate(ctx => ctx.DrawText(new RichTextOptions(font) { KerningMode = KerningMode.Auto, Origin = new PointF(x, y) }, "PLAYER NAME", color));

If I try changing the KerningMode to None, the Y and E get better (they no longer touch) but then there is a significantly larger gap between the A and Y. How can I get this closer to what renders in the SVG? Please keep in mind that in my application/SVG the text and font are dynamic - so I'm looking for a one size fits all solution
![image](https://github.com/SixLabors/Fonts/assets/433538/9206ca8a-a7fa-45bd-8186-690fcadd5c69)

**Comment by avisra** [comment]
For the record, even skia sharp generates the A and Y spaced apart like the above. I was only able to fix it in SkiaSharp by using the richtext kit plugin here: https://www.toptensoftware.com/richtextkit/

I'm not sure this is helpful at all... but maybe there is something we could borrow from their plugin for glyph spacing?

And as far as the browser goes with the SVG, if I set font-kerning: none on the element - then it matches that SixLabors is producing. But if I set font-kerning: auto, it produces what you see in the blue screenshot at the top - which is something I have not been able to get SixLabors to produce

**Comment by tocsoft** [comment]
It seems like the kerning is being applied to the wrong character pair, looking at how all the character are spaced in both the 'with kerning' and 'without kerning' images when kerning is enable, the `E` (and subsequent letter) have been moved to the left rather than the `Y` moving to the right, as the AY spacing is consistent in both images.

Not that I know what the fix is or what could be causing it but it seems to effect all kerning in the font, kerning pairs targeting capital `V` are also effected and not just with `A` but things like `a`, `o`, `L` also trigger the behaviour.

Also this behaviour doesn't trigger with all fonts, Open Sans for example isn't affected so it must be a font specific thing where there is an in font setting that is triggering this issue.

**Reply by avisra** [reply]
I would be willing to work on this if anyone had a hint of where I should start. I have a lot of experience with ImageSharp - but this is my first exposure to fonts.

**Reply by JimBobSquarePants** [reply]
There's a couple of different online tools I use to compare the output. 

FontKit
https://fontkit-demo.vercel.app/

Crowbar (Harfbuzz)
https://www.corvelsoftware.co.uk/crowbar/

The latter is extremely useful when debugging substitution and positioning

Both get this correct. 

Looking at the reference source for FontKit it becomes obvious that we have been applying the kerning offset to the wrong glyph.

![image](https://github.com/SixLabors/Fonts/assets/385879/6be5d4d4-8d04-4072-a060-c9b5e15008bd)

https://github.com/SixLabors/Fonts/blob/e546be3ff828a82cab095a58ceaa5f9063f8a133/src/SixLabors.Fonts/Tables/General/Kern/KerningTable.cs#L58-L73

Most fonts do the kerning via the GPOS tables (using Kern and VKern feature tags) so we hadn't picked up on the issue in our existing tests however this font relies on the KERN table. 

A quick update to the table and calling methods and we can get the correct result. 

![PLAYER NAME](https://github.com/SixLabors/Fonts/assets/385879/9ccf3b5c-c58f-4775-926d-c25fa5c7db9c)

Unfortunately, the fix will contain a breaking change because there are public calling methods that implement the wrong semantics. Unavoidable though.

**Reply by JimBobSquarePants** [reply]
PR here. https://github.com/SixLabors/Fonts/pull/405

**Reply by avisra** [reply]
Holy moly. I was not expecting to wake up this morning to a fix. Thanks @JimBobSquarePants!!

**Reply by JimBobSquarePants** [reply, marked as answer]
Even better… The release is on NuGet now. 😀
